pkg/payload/task_graph.go

```diff
@@ -437,169 +435,117 @@ func RunGraph(ctx context.Context, graph *TaskGraph, maxParallelism int, fn func
-				klog.V(4).Infof("Stopped graph walker due to cancel")
-				return
-			}
-			if finished.success {
-				visited[finished.index] = nodeComplete
-			} else {
-				visited[finished.index] = nodeFailed
-			}
-			remaining--
-			inflight--
-		}
+	// Tasks go out to the workers via workCh, and results come brack
+	// from the workers via resultCh.
+	workCh := make(chan runTasks, maxParallelism)
+	defer close(workCh)
 
-		// take everything remaining and process in order
-		var unreachable []*Task
-		for i := 0; i < len(visited); i++ {
-			if visited[i] == nodeNotVisited && canVisit(graph.Nodes[i]) {
-				unreachable = append(unreachable, graph.Nodes[i].Tasks...)
-			}
-		}
-		if len(unreachable) > 0 {
-			sort.Slice(unreachable, func(i, j int) bool {
-				a, b := unreachable[i], unreachable[j]
-				return a.Index < b.Index
-			})
-			workCh <- runTasks{index: -1, tasks: unreachable}
-			klog.V(4).Infof("Waiting for last tasks")
-			<-completeCh
-		}
-		klog.V(4).Infof("No more work")
-	}()
+	resultCh := make(chan taskStatus, maxParallelism)
+	defer close(resultCh)
 
-	errCh := make(chan error, maxParallelism)
 	wg := sync.WaitGroup{}
 	if maxParallelism < 1 {
 		maxParallelism = 1
 	}
 	for i := 0; i < maxParallelism; i++ {
 		wg.Add(1)
-		go func(job int) {
+		go func(ctx context.Context, job int) {
 			defer utilruntime.HandleCrash()
 			defer wg.Done()
 			for {
 				select {
-				case <-nestedCtx.Done():
-					klog.V(4).Infof("Canceled worker %d", job)
+				case <-ctx.Done():
+					klog.V(4).Infof("Canceled worker %d while waiting for work", job)
 					return
-				case runTask, ok := <-workCh:
-					if !ok {
-						klog.V(4).Infof("No more work for %d", job)
-						return
-					}
+				case runTask := <-workCh:
 					klog.V(4).Infof("Running %d on %d", runTask.index, job)
-					err := fn(nestedCtx, runTask.tasks)
-					completeCh <- taskStatus{index: runTask.index, success: err == nil}
-					if err != nil {
-						errCh <- err
-					}
+					err := fn(ctx, runTask.tasks)
+					resultCh <- taskStatus{index: runTask.index, error: err}
 				}
 			}
-		}(i)
+		}(nestedCtx, i)
 	}
-	go func() {
-		klog.V(4).Infof("Waiting for workers to complete")
-		wg.Wait()
-		klog.V(4).Infof("Workers finished")
-		close(errCh)
-	}()
+
+	var inflight int
+	nextNode := getNextNode()
+	done := false
+	for !done {
+		switch {
+		case ctx.Err() == nil && nextNode >= 0: // push a task or collect a result
+			select {
+			case workCh <- runTasks{index: nextNode, tasks: graph.Nodes[nextNode].Tasks}:
+				submitted[nextNode] = true
+				inflight++
+			case result := <-resultCh:
+				results[result.index] = &result
+				inflight--
+			case <-ctx.Done():
+			}
+		case inflight > 0: // no work available to push; collect results
+			result := <-resultCh
+			results[result.index] = &result
+			inflight--
+		default: // no work to push and nothing in flight.  We're done
+			done = true
+		}
+		if !done {
+			nextNode = getNextNode()
+		}
+	}
+
+	cancelFn()
```

Comment:
the runner is 
- waiting on ctx
- waiting on work from workCh

the also passes along the ctx to the syn task fn.

maybe what we should do is,
the runner only loops on the workCh for work, and passes along the ctx to task sync fn so that we can terminate in progress work.
here we should just close workCh as there is no longer any work left to done.

that simplifies/compartmentalizes what's used for what.

Reply:
From [the Go spec][1]:

> Sending to or closing a closed channel causes a run-time panic.

So if we manually call `close(workCh)` [here][2], then [this deferred call][3] will panic, right?  Maybe Go ignores panics from deferred functions, but still, I personally prefer having the worker watching both `nestedCtx` and `workCh`.

[1]: https://golang.org/ref/spec#Close
[2]: https://github.com/openshift/cluster-version-operator/commit/f6271d01756b8c744c1cbaf57dd9ea79ff8b836e#diff-bf8962ebeca893755f0b6fbd9b0252eaR507
[3]: https://github.com/openshift/cluster-version-operator/commit/f6271d01756b8c744c1cbaf57dd9ea79ff8b836e#diff-bf8962ebeca893755f0b6fbd9b0252eaR463